lib/spack/spack/compiler.py

```diff
@@ -174,6 +176,103 @@ def check(exe):
     def version(self):
         return self.spec.version
 
+    # This property should be overridden in the compiler subclass if
+    # a verbose flag is available
+    @property
+    def verbose_flag(self):
+        # If it is not overridden, assume it is not supported
+        pass
+
+    _LINKER_LINE = re.compile(
+        r'^( *|.*[/\\])(link|ld|([^/\\]+-)?ld|collect2)[^/\\]*( |$)')
+    _LINKER_LINE_IGNORE = re.compile(
+        r'(collect2 version|^[A-Za-z0-9_]+=|/ldfe )')
+    _LINK_DIR_ARG = re.compile(r'^-L(.:)?[/\\]')
+    _LIBPATH_ARG = re.compile(r'^[-/](LIBPATH|libpath):')
+
+    def _determine_implicit_link_paths(self):
+        try:
+            compilers = [self.cc, self.cxx, self.f77, self.fc]
+            first_compiler = filter(None, compilers)[0]
+        except IndexError:
+            # No compilers available.
+            return []
+
+        try:
+            tmpdir = tempfile.mkdtemp(prefix='spack-implicit-link-info')
+            fout = os.path.join(tmpdir, 'output')
+            fin = os.path.join(tmpdir, 'main.c')
+
+            with open(fin, 'w+') as csource:
+                csource.write(
+                    'int main(int argc, char* argv[]) { '
+                    '(void)argc; (void)argv; return 0; }\n')
+                csource.flush()
+            output = Executable(first_compiler)(self.verbose_flag, fin, '-o',
+                                                fout, output=str, error=str)
+
+            lib_search_paths = False
+            raw_link_dirs = []
+            tty.debug('parsing implicit link info')
+            for line in output.splitlines():
+                if lib_search_paths:
+                    if line.startswith('\t'):
+                        raw_link_dirs.append(line[1:])
+                        continue
+                    else:
+                        lib_search_paths = False
+                elif line.startswith('Library search paths:'):
+                    lib_search_paths = True
+
+                if not self._LINKER_LINE.match(line):
+                    pass
+                if self._LINKER_LINE_IGNORE.match(line):
+                    pass
+                tty.debug('linker line: %s' % line)
+
+                next_arg = False
+                for arg in line.split():
+                    if arg in ('-L', '-Y'):
+                        next_arg = True
+                        continue
+
+                    if next_arg:
+                        raw_link_dirs.append(arg)
+                        next_arg = False
+                        continue
+
+                    if self._LINK_DIR_ARG.match(arg):
+                        tty.debug('linkdir: %s' % arg)
+                        raw_link_dirs.append(arg[2:])
```

Comment:
instead of remembering the index here and on the next condition, can you extract the linkdir and libpath with parens?  That will be cleaner if the expressions change.